tests/onebox.go

```diff
@@ -271,112 +290,56 @@ func (c *temporalImpl) Stop() error {
+	hostMap := maps.Clone(c.hostsByService)
+	hosts := hostMap[serviceName]
+	hosts.Self = self
+	hostMap[serviceName] = hosts
+	return hostMap
 }
 
-func (c *temporalImpl) FrontendGRPCAddress() string {
-	switch c.clusterNo {
-	case 0:
-		return "127.0.0.1:7134"
-	case 1:
-		return "127.0.0.1:8134"
-	case 2:
-		return "127.0.0.1:9134"
-	case 3:
-		return "127.0.0.1:10134"
-	default:
-		return "127.0.0.1:7134"
+func (c *temporalImpl) makeGRPCAddresses(num, port int) []string {
+	hosts := make([]string, num)
+	for i := range hosts {
+		hosts[i] = fmt.Sprintf("127.0.%d.%d:%d", c.clusterNo, i+1, port)
 	}
+	return hosts
 }
 
-func (c *temporalImpl) FrontendHTTPAddress() string {
-	host, port := c.FrontendHTTPHostPort()
-	return net.JoinHostPort(host, strconv.Itoa(port))
+func (c *temporalImpl) FrontendGRPCAddresses() []string {
+	return c.makeGRPCAddresses(c.frontendConfig.NumFrontendHosts, frontendPort)
 }
 
-func (c *temporalImpl) FrontendHTTPHostPort() (string, int) {
-	if host, port, err := net.SplitHostPort(c.FrontendGRPCAddress()); err != nil {
+func (c *temporalImpl) FrontendHTTPAddress() string {
+	// randomize like a load balancer would
+	addrs := c.FrontendGRPCAddresses()
+	addr := addrs[rand.Intn(len(addrs))]
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
 		panic(fmt.Errorf("Invalid gRPC frontend address: %w", err))
-	} else if portNum, err := strconv.Atoi(port); err != nil {
-		panic(fmt.Errorf("Invalid gRPC frontend port: %w", err))
-	} else {
-		return host, portNum + 10
 	}
+	return net.JoinHostPort(host, strconv.Itoa(frontendHTTPPort))
 }
 
 func (c *temporalImpl) HistoryServiceAddresses() []string {
-	var hosts []string
-	var startPort int
-	switch c.clusterNo {
-	case 0:
-		startPort = 7231
-	case 1:
-		startPort = 8231
-	case 2:
-		startPort = 9231
-	case 3:
-		startPort = 10231
-	default:
-		startPort = 7231
-	}
-	for i := 0; i < c.historyConfig.NumHistoryHosts; i++ {
-		port := startPort + i
-		hosts = append(hosts, fmt.Sprintf("127.0.0.1:%v", port))
-	}
-
-	c.logger.Info("History hosts", tag.Addresses(hosts))
-	return hosts
+	return c.makeGRPCAddresses(c.historyConfig.NumHistoryHosts, historyPort)
 }
 
-func (c *temporalImpl) MatchingGRPCServiceAddress() string {
-	switch c.clusterNo {
-	case 0:
-		return "127.0.0.1:7136"
-	case 1:
-		return "127.0.0.1:8136"
-	case 2:
-		return "127.0.0.1:9136"
-	case 3:
-		return "127.0.0.1:10136"
-	default:
-		return "127.0.0.1:7136"
-	}
+func (c *temporalImpl) MatchingServiceAddresses() []string {
+	return c.makeGRPCAddresses(c.matchingConfig.NumMatchingHosts, matchingPort)
 }
 
-func (c *temporalImpl) WorkerGRPCServiceAddress() string {
-	switch c.clusterNo {
-	case 0:
-		return "127.0.0.1:7138"
-	case 1:
-		return "127.0.0.1:8138"
-	case 2:
-		return "127.0.0.1:9138"
-	case 3:
-		return "127.0.0.1:10138"
-	default:
-		return "127.0.0.1:7138"
-	}
+func (c *temporalImpl) WorkerServiceAddresses() []string {
+	return c.makeGRPCAddresses(c.workerConfig.NumWorkers, workerPort)
```

Comment:
Why to have this at all?

Reply:
They need unique identities so the workers can tell themselves apart, for assigning per-ns workers or whatever else might use hostinfo/serviceresolver